tools/flowc/const_eval.flow

```diff
@@ -0,0 +1,861 @@
+	make_err = \msg, pos -> { on_err(msg, pos); None(); }
+	switch (call.op) {
+		FcOrPrim(): {
+			a1 = cast(args[0]: FiExp -> FiBool);
+			a2 = cast(args[1]: FiExp -> FiBool);
+			Some(FiBool(a1.b || a2.b, a1.start));
+		}
+		FcAndPrim(): {
+			a1 = cast(args[0]: FiExp -> FiBool);
+			a2 = cast(args[1]: FiExp -> FiBool);
+			Some(FiBool(a1.b && a2.b, a1.start));
+		}
+		FcNotPrim(): {
+			a1 = cast(args[0]: FiExp -> FiBool);
+			Some(FiBool(!a1.b, a1.start));
+		}
+		FcEqPrim(): {
+			switch (args[0]) {
+				FiLambda(__,__,__,__): None();
+				default: Some(FiBool(fiConstExpCompare(args[0], args[1]) == 0, args[0].start));
+			}
+		}
+		FcNePrim(): {
+			switch (args[0]) {
+				FiLambda(__,__,__,__): None();
+				default: Some(FiBool(fiConstExpCompare(args[0], args[1]) != 0, args[0].start));
+			}
+		}
+		FcLePrim(): {
+			switch (args[0]) {
+				FiLambda(__,__,__,__): None();
+				default: Some(FiBool(fiConstExpCompare(args[0], args[1]) <= 0, args[0].start));
+			}
+		}
+		FcGePrim(): {
+			switch (args[0]) {
+				FiLambda(__,__,__,__): None();
+				default: Some(FiBool(fiConstExpCompare(args[0], args[1]) >= 0, args[0].start));
+			}
+		}
+		FcLtPrim(): {
+			switch (args[0]) {
+				FiLambda(__,__,__,__): None();
+				default: Some(FiBool(fiConstExpCompare(args[0], args[1]) < 0, args[0].start));
+			}
+		}
+		FcGtPrim(): {
+			switch (args[0]) {
+				FiLambda(__,__,__,__): None();
+				default: Some(FiBool(fiConstExpCompare(args[0], args[1]) > 0, args[0].start));
+			}
+		}
+		FcPlusPrim(): {
+			switch (args[0]) {
+				FiInt(__,__): {
+					a1 = cast(args[0]: FiExp -> FiInt);
+					a2 = cast(args[1]: FiExp -> FiInt);
+					Some(FiInt(a1.i + a2.i, a1.start));
+				}
+				FiDouble(__,__): {
+					a1 = cast(args[0]: FiExp -> FiDouble);
+					a2 = cast(args[1]: FiExp -> FiDouble);
+					Some(FiDouble(a1.d + a2.d, a1.start));
+				}
+				FiString(__,__): {
+					a1 = cast(args[0]: FiExp -> FiString);
+					a2 = cast(args[1]: FiExp -> FiString);
+					Some(FiString(a1.s + a2.s, a1.start));
+				}
+				default: None();
+			}
+		}
+		FcMinusPrim(): {
+			switch (args[0]) {
+				FiInt(__,__): {
+					a1 = cast(args[0]: FiExp -> FiInt);
+					a2 = cast(args[1]: FiExp -> FiInt);
+					Some(FiInt(a1.i - a2.i, a1.start));
+				}
+				FiDouble(__,__): {
+					a1 = cast(args[0]: FiExp -> FiDouble);
+					a2 = cast(args[1]: FiExp -> FiDouble);
+					Some(FiDouble(a1.d - a2.d, a1.start));
+				}
+				default: None();
+			}
+		}
+		FcMulPrim(): {
+			switch (args[0]) {
+				FiInt(__,__): {
+					a1 = cast(args[0]: FiExp -> FiInt);
+					a2 = cast(args[1]: FiExp -> FiInt);
+					Some(FiInt(a1.i * a2.i, a1.start));
+				}
+				FiDouble(__,__): {
+					a1 = cast(args[0]: FiExp -> FiDouble);
+					a2 = cast(args[1]: FiExp -> FiDouble);
+					Some(FiDouble(a1.d * a2.d, a1.start));
+				}
+				default: None();
+			}
+		}
+		FcDivPrim(): {
+			switch (args[0]) {
+				FiInt(__,__): {
+					a1 = cast(args[0]: FiExp -> FiInt);
```

Comment:
If you bound args[0] to a name, the cast is not necessary

Reply:
Fixed.